Implement Linux support

Jan200101 opened the pull request:
late followup to #55
closes #43 

- Windows specific logic has been gated behind an if case
- APPDATA has been globalized in many cases to reduce calls to os.getenv as well as to use $HOME/.config on non NT systems
- a dependency on a posix compatibility layer for winreg has been added for non Windows systems
  - no need to rewrite huge chunks of the code
  - has been written specificially for this purpose

Things to implement before this can be merged:
- [x] game detection
- [ ] Steam Username detection
  - there seems to be no reliable way to detect the current steam username so we just use the users name so its still unique. Most people won't have more than 1 account either way
- [ ] building on Linux
  - building is a huge mess and its best to not do this on Linux at all since distros all have ship different libraries and deal with them differently

Jan200101 commented on TF2 Rich Presence/build.py:
```diff
@@ -477,7 +477,7 @@ def copy_dir(source, target):
     if sys.platform == 'win32':
         subprocess.run(f'xcopy \"{source}\" \"{target}{os.path.sep}\" /E /Q')
```
quick question about this: what is the purpose of calling `xcopy` instead of just using `shutil.copy`?
Are there any special properties that I am not aware of

Kataiser replied:
IIRC it was a lot faster

Jan200101 commented on requirements.txt:
```diff
@@ -12,4 +12,5 @@ requests-futures==1.0.0
-vdf==3.4
\ No newline at end of file
+vdf==3.4
+unixreg; sys_platform != 'win32'
```
If having another external dependency is undesirable its possible to include the code in this repo

Kataiser replied:
It's fine if it's not big

Jan200101 replied:
For reference
the source code of unixreg is available on PyPi and Github

https://pypi.org/project/unixreg/
https://github.com/Jan200101/unixreg

Jan200101 commented on TF2 Rich Presence/utils.py:
```diff
@@ -9,6 +9,10 @@
+if os.name == "nt":
+    APPDATA = os.getenv('APPDATA')
+else:
+    APPDATA = os.path.join(os.getenv("HOME"), ".config")
```
needs to be improved to respect XDG_CONFIG_HOME

Jan200101 commented on TF2 Rich Presence/settings.py:
```diff
@@ -28,7 +31,7 @@ def access_registry(save: Optional[dict] = None) -> Optional[dict]:
     try:
         reg_key_data: dict = json.loads(winreg.QueryValue(reg_key, 'Settings'))
-    except FileNotFoundError:  # means that the key hasn't been initialized
+    except (FileNotFoundError, json.JSONDecodeError):  # means that the key hasn't been initialized
```
I could not find what is suppose to throw FileNotFoundError so I simply fell back to catching JSONDecodeError

Kataiser replied:
Raised by winreg.QueryValue() I believe

Jan200101 replied:
Checked this on Windows and it does indeed throw FileNotFound.

Reverted this change and implemented throwing this exception in the compatibility library

Jan200101 commented on TF2 Rich Presence/localization.py:
```diff
@@ -119,7 +118,7 @@ def detect_system_language(log: logger.Log):
     if not db['has_asked_language']:
         language_codes: dict = {read_localization_files()[lang]['code']: lang for lang in langs[1:]}
-        system_locale: str = locale.windows_locale[ctypes.windll.kernel32.GetUserDefaultUILanguage()]
+        system_locale: str = locale.getlocale()[0]
```
is `GetUserDefaultUILanguage` more accurate than `getlocale` ?

Kataiser replied:
IDK I probably copied this line from somewhere

Kataiser commented:
Yo this is amazing! I'd love to help see this through. I know absolutely nothing about releasing on Linux so you'd have to help me a bit with that, but I'd maintain a Linux build for sure

Jan200101 committed:
Implement Linux support
- disable building on Linux (TODO)
- add winreg compatibility layer to allow reusing existing code without having to write an new system
- replace ctypes usage available python replacement
- gate windows logic behind os.name == "nt" checks (TODO compare how this behaves on cygwin and msys)
- update hl2 linux executable name to use the script instead
- replace hardcoded usage of "*.exe" with indexing of the executables list

Jan200101 commented:
![image](https://user-images.githubusercontent.com/15076013/150643187-39138204-f174-461f-80fe-2bffd440b8ed.png)
its now in a working state again

Kataiser commented:
Nice!
